# Data Aggregation and Group Operations

## Grouping
By “**group by**” we are referring to a process involving one or more of the following steps:

- Splitting the data into groups based on some criteria

- Applying a function to each group independently

- Combining the results into a data structure

![](images/group_by.jpg)

**Illustration of a group aggregation**

In [3]:
import numpy as np

import pandas as pd

from pandas_profiling import ProfileReport

## Example 1

Consider `penguins` DataFrame, what is the mean `bill_length`(mm) of each of the species.

In [4]:
penguins = pd.read_csv("datasets/penguins.csv")

In [5]:
penguins.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181,3750,male
1,Adelie,Torgersen,39.5,17.4,186,3800,female
2,Adelie,Torgersen,40.3,18.0,195,3250,female
3,Adelie,Torgersen,36.7,19.3,193,3450,female
4,Adelie,Torgersen,39.3,20.6,190,3650,male


We call the `.groupby()` on the DataFrame and with a column (a Series) to groupby with. We then apply some operation to
each of the groups. For example, to compute group means we can call the groupby’s mean method:

In [6]:
penguins.groupby("species").mean()

,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
species,,,,
Adelie,38.800000,18.342667,190.026667,3705.500000
Chinstrap,48.833824,18.420588,195.823529,3733.088235
Gentoo,47.568067,14.996639,217.235294,5092.436975


We can now see the mean of each continuous variable for each of the species.

The mean bill_length of Adelie, Chinstrap, and Gentoo are 38.800000, 48.833824	, and 47.568067 respectively.

We can actually use:

In [181]:
penguins.groupby(["species"])["bill_length_mm"].mean()

species
Adelie       38.800000
Chinstrap    48.833824
Gentoo       47.568067
Name: bill_length_mm, dtype: float64

To get the same results with only mean of the bill_length_mm for each specie

## More than one grouping

Groupby also allows for more than 1 grouping. For example, what is the sum of the bill_depth (mm) for the combination of species and sex of Penguins?

In [182]:
penguins.groupby(["species", "sex"])["bill_length_mm"].sum()

species    sex   
Adelie     female    2791.7
           male      3028.3
Chinstrap  female    1583.5
           male      1737.2
Gentoo     female    2642.7
           male      3017.9
Name: bill_length_mm, dtype: float64

# Cross-Tabulations (Crosstab)

Contingency tables also called crosstabs or two-way tables are used in statistics to present categorical data in terms of frequency counts. In other words, it is a special type of frequency distribution table, where two variables are shown simultaneously. More precisely, an r×c contingency table shows the observed frequency of two variables, the observed frequencies of which are arranged into r rows and c columns. The intersection of a row and a column of a contingency table is called a cell.

## How to do crosstabulation?

We use `pd.crosstab()` attribute to do crosstabulation and `pd.crosstab()` contains the following:

`pd.crosstab(index, columns, values=None, rownames=None, colnames=None, aggfunc=None, margins=False, margins_name='All', dropna=True, normalize=False)` 

For more information about `pd.crosstab()`, check this [link](https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html).

Some major parameters are:

* `index`: Series- values to group by in the rows

* `column`: Series - values to group by in the columns

* `margins` bool (True/False), default False: Add row/column margins (subtotals)

In [185]:
pd.crosstab(index = penguins["species"], columns = penguins["sex"], margins = True)

sex,female,male,All
species,,,
Adelie,75,75,150
Chinstrap,34,34,68
Gentoo,58,61,119
All,167,170,337


As you can see, there are 58 female and 61 male Gentoo Penguins

## Example 2

Create a crosstabulation table using index for species, Island and sex for columns

In [7]:
pd.crosstab(index = penguins["species"], columns = penguins["island"], margins=True)

island,Biscoe,Dream,Torgersen,All
species,,,,
Adelie,44,55,51,150
Chinstrap,0,68,0,68
Gentoo,119,0,0,119
All,163,123,51,337
